client-v2/src/selectors/frontsSelectors.js

```diff
@@ -4,64 +4,68 @@ import { createSelector } from 'reselect';
 import { breakingNewsFrontId } from '../constants/fronts';
 
 import type {
-  FrontConfig,
-  FrontDetail,
-  FrontsClientConfig,
   Front,
-  ConfigCollection
+  ConfigCollection,
+  FrontDetailFull,
+  FrontsWithIds,
+  FrontsByPriority
 } from '../types/FrontsConfig';
 
 import type { State } from '../types/State';
 
-const frontsSelector = (state: State): Front => state.frontsConfig.fronts;
-const collectionsSelector = (state: State): ConfigCollection =>
-  state.frontsConfig.collections;
+const getFrontsConfig = (state: State) => state.frontsConfig || {};
 
-const prioritySelector = (state: State, priority: string) => priority;
+const rawFrontsSelector = (state: State): Front =>
+  getFrontsConfig(state).fronts || {};
 
-const frontsIdSelector = createSelector([frontsSelector], fronts => {
+const getFronts = createSelector(
+  [rawFrontsSelector],
+  (fronts: Front): FrontsWithIds =>
+    Object.keys(fronts)
+      .map((id): FrontDetailFull => ({
+        id,
+        ...fronts[id],
+        priority: fronts[id].priority || 'editorial'
+      }))
+      .reduce(
+        (acc: FrontsWithIds, front: FrontDetailFull): FrontsWithIds => ({
+          ...acc,
+          [front.id]: front
+        }),
+        {}
+      )
+);
+
+const getFrontsByPriority = createSelector(
+  [getFronts],
+  (fronts: FrontsWithIds): FrontsByPriority =>
+    Object.keys(fronts).reduce(
+      (acc: FrontsByPriority, id): FrontsByPriority => {
+        const front = fronts[id];
+        return {
+          ...acc,
+          [front.priority]: [...(acc[front.priority] || []), fronts[id]]
+        };
+      },
+      {}
+    )
+);
+
+const getFrontsWithPriority = (
+  state: State,
+  priority: string
+): FrontDetailFull[] => getFrontsByPriority(state)[priority] || [];
```

Comment:
Users are unlikely to do a lot of switching around different priorities so sorting all fronts by priorities in the selectors might be too much complexity but also it is a nice thing to have

Reply:
So the reason I did this is that it means the selectors don't break when we pass a different props object. It's annoyingly subtle but basically, when we were passing a props object before to the selectors, even if the `priority` key was the same, if we passed an object with a different reference (i.e. when another prop on the container that was using it changed) then `reselect` would recompute the selector because it does shallow equality checking. This might not make sense so maybe we can talk about it as it's not intuitive at all.

Reply:
☝️  This is the reason that I've refactored a few of the other selectors.

Reply:
Ok, I think I get it. So if we are passing in component props to selectors, we should separate out the logic that uses the prop from all other data manipulation we are doing.